src/java/org/apache/cassandra/db/counters/CachedCounterLockManager.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.counters;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.StampedLock;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+
+/**
+ * Implemetation of {@link CounterLockManager} that uses a cache of locks.
+ * Note: this implemetation tries to reduce the chance of having two counters lock each other, but as the counters
+ * are identified by the hash of the primary key of the row, it is still possible to
+ * have some cross-counter contention for counters with different primary keys but the same hash.
+ *
+ * This code is copied from
+ * <a href="https://github.com/diennea/herddb/blob/master/herddb-utils/src/main/java/herddb/utils/LocalLockManager.java">LocalLockManager</a> from the HerdDB project (Apache 2 licensed).
+ *
+ */
+public class CachedCounterLockManager implements CounterLockManager
+{
+    private final static int EXPECTED_CONCURRENCY = DatabaseDescriptor.getConcurrentCounterWriters() * 16;
+    private final ConcurrentMap<Integer, RefCountedStampedLock> locks = new ConcurrentHashMap<>(EXPECTED_CONCURRENCY);
```

Comment:
the `makeLockForKey` function relies on the CHM guarantee that the remapping function is run only once per `compute`, and that it is run atomically.
Please add a comment to that effect and change the type of the `locks` to CHM. 
It is not enough that `locks` fulfill `ConcurrentMap` contract

Reply:
I added a comment, good point